File: matrix_operations.py
import numpy as np


class MatrixOperations:

    # ================= SAFE ADD =================
    def add(self, A, B):
        if A.shape != B.shape:
            raise ValueError("Addition requires same dimensions")
        return np.add(A, B)

    # ================= SAFE MULTIPLY =================
    def multiply(self, A, B):
        if A.shape[1] != B.shape[0]:
            raise ValueError("Invalid multiplication dimensions")
        return np.matmul(A, B)

    # ================= TRANSPOSE =================
    def transpose(self, A):
        return np.transpose(A)

    # ================= DETERMINANT =================
    def determinant(self, A):
        if A.shape[0] != A.shape[1]:
            raise ValueError("Determinant requires square matrix")
        return np.linalg.det(A)

    # ================= INVERSE =================
    def inverse(self, A):
        if A.shape[0] != A.shape[1]:
            raise ValueError("Inverse requires square matrix")
        return np.linalg.inv(A)

    # ================= ADJOINT =================
    def adjoint(self, A):
        if A.shape[0] != A.shape[1]:
            raise ValueError("Adjoint requires square matrix")

        det = np.linalg.det(A)

        if np.isclose(det, 0):
            raise ValueError("Matrix is singular (no adjoint via inverse method)")

        inv = np.linalg.inv(A)
        return det * inv

    # ================= POWER =================
    def power(self, A, n):
        if A.shape[0] != A.shape[1]:
            raise ValueError("Power requires square matrix")
        return np.linalg.matrix_power(A, n)

    # ================= SOLVE AX = B =================
    def solve_equations(self, A, B):
        if A.shape[0] != A.shape[1]:
            raise ValueError("A must be square for solving equations")

        if A.shape[0] != B.shape[0]:
            raise ValueError("Row mismatch between A and B")

        return np.linalg.solve(A, B)

File: Matlab/MatrixOperations.m
classdef MatrixOperations

    methods

        function C = add(~, A, B)
            if ~isequal(size(A), size(B))
                error('Addition requires same dimensions');
            end
            C = A + B;
        end

        function C = multiply(~, A, B)
            if size(A,2) ~= size(B,1)
                error('Invalid multiplication dimensions');
            end
            C = A * B;
        end

        % TRANSPOSE
        function T = transpose(~, A)
            T = A.';   % Pure transpose
        end

        function d = determinant(~, A)
            if size(A,1) ~= size(A,2)
                error('Determinant requires square matrix');
            end
            d = det(A);
        end

        % INVERSE
        function invA = inverse(~, A)
            if size(A,1) ~= size(A,2)
                error('Inverse requires square matrix');
            end

            if abs(det(A)) < 1e-12
                error('Matrix is singular and cannot be inverted');
            end

            invA = inv(A);
        end

        % ADJOINT
        function adjA = adjoint(~, A)
            if size(A,1) ~= size(A,2)
                error('Adjoint requires square matrix');
            end

            d = det(A);

            if abs(d) < 1e-12
                error('Singular matrix');
            end

            adjA = d * inv(A);

            % Remove floating point errors
            adjA(abs(adjA) < 1e-10) = 0;
        end

        function P = power(~, A, n)
            if size(A,1) ~= size(A,2)
                error('Power requires square matrix');
            end
            P = A^n;
        end

        function X = solve_equations(~, A, B)
            if size(A,1) ~= size(A,2)
                error('A must be square');
            end
            X = A \ B;
        end
    end
end
